// @predy-js/player/dist/editor.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var renderInterface = require('@predy-js/render-interface');

const END_BEHAVIOR_DESTROY = 0;
const END_BEHAVIOR_PAUSE = 1;
const END_BEHAVIOR_FORWARD = 2;
const END_BEHAVIOR_PAUSE_AND_DESTROY = 3;
const END_BEHAVIOR_FREEZE = 4;
const END_BEHAVIOR_RESTART = 5;
const END_BEHAVIOR_DESTROY_CHILDREN = 6;
const CAMERA_CLIP_MODE_VERTICAL = 1;
const CAMERA_CLIP_MODE_NORMAL = 0;

var ItemEndBehavior;
(function (ItemEndBehavior) {
    ItemEndBehavior[ItemEndBehavior["destroy"] = END_BEHAVIOR_DESTROY] = "destroy";
    ItemEndBehavior[ItemEndBehavior["loop"] = END_BEHAVIOR_RESTART] = "loop";
    ItemEndBehavior[ItemEndBehavior["freeze"] = END_BEHAVIOR_FREEZE] = "freeze";
    ItemEndBehavior[ItemEndBehavior["destroyChildren"] = END_BEHAVIOR_DESTROY_CHILDREN] = "destroyChildren";
})(ItemEndBehavior || (ItemEndBehavior = {}));

var SpriteInteractionBehavior;
(function (SpriteInteractionBehavior) {
    SpriteInteractionBehavior[SpriteInteractionBehavior["none"] = 0] = "none";
})(SpriteInteractionBehavior || (SpriteInteractionBehavior = {}));

//import type { DistortionFilterParams } from './filter-item';
/**
 * 粒子交互行为
 */
var ParticleInteractionBehavior;
(function (ParticleInteractionBehavior) {
    /**
       * 无
       */
    ParticleInteractionBehavior[ParticleInteractionBehavior["none"] = 0] = "none";
    /**
       * 移出粒子
       */
    ParticleInteractionBehavior[ParticleInteractionBehavior["removeParticle"] = 1] = "removeParticle";
})(ParticleInteractionBehavior || (ParticleInteractionBehavior = {}));

var ShapeArcMode;
(function (ShapeArcMode) {
    /**
     * 随机
     */
    ShapeArcMode[ShapeArcMode["RANDOM"] = 0] = "RANDOM";
    /**
     * 单向循环
     */
    ShapeArcMode[ShapeArcMode["UNIDIRECTIONAL_CYCLE"] = 1] = "UNIDIRECTIONAL_CYCLE";
    /**
     * 双向循环
     */
    ShapeArcMode[ShapeArcMode["BIDIRECTIONAL_CYCLE"] = 2] = "BIDIRECTIONAL_CYCLE";
    /**
     * 均匀爆发
     */
    ShapeArcMode[ShapeArcMode["UNIFORM_BURST"] = 3] = "UNIFORM_BURST";
})(ShapeArcMode || (ShapeArcMode = {}));

var VideoRendererSplitAlpha;
(function (VideoRendererSplitAlpha) {
    VideoRendererSplitAlpha[VideoRendererSplitAlpha["none"] = 2] = "none";
    VideoRendererSplitAlpha[VideoRendererSplitAlpha["c2a"] = 0] = "c2a";
    VideoRendererSplitAlpha[VideoRendererSplitAlpha["a2c"] = 1] = "a2c"; // alpha 在左边，color 在右边
})(VideoRendererSplitAlpha || (VideoRendererSplitAlpha = {}));

var LightType;
(function (LightType) {
    LightType["point"] = "point";
    LightType["spot"] = "spot";
    LightType["directional"] = "directional";
    LightType["ambient"] = "ambient";
})(LightType || (LightType = {}));

var ModelBoundingType;
(function (ModelBoundingType) {
    ModelBoundingType[ModelBoundingType["box"] = 2] = "box";
    ModelBoundingType[ModelBoundingType["sphere"] = 3] = "sphere";
})(ModelBoundingType || (ModelBoundingType = {}));

var TreeAnimPathType;
(function (TreeAnimPathType) {
    TreeAnimPathType[TreeAnimPathType["translation"] = 1] = "translation";
    TreeAnimPathType[TreeAnimPathType["rotation"] = 2] = "rotation";
    TreeAnimPathType[TreeAnimPathType["scale"] = 3] = "scale";
    TreeAnimPathType[TreeAnimPathType["weights"] = 4] = "weights";
})(TreeAnimPathType || (TreeAnimPathType = {}));
var TreeAnimInterpType;
(function (TreeAnimInterpType) {
    TreeAnimInterpType[TreeAnimInterpType["linear"] = 1] = "linear";
    TreeAnimInterpType[TreeAnimInterpType["step"] = 2] = "step";
    TreeAnimInterpType[TreeAnimInterpType["cubicSpline"] = 3] = "cubicSpline";
    TreeAnimInterpType[TreeAnimInterpType["LINEAR"] = 1] = "LINEAR";
    TreeAnimInterpType[TreeAnimInterpType["STEP"] = 2] = "STEP";
    TreeAnimInterpType[TreeAnimInterpType["CUBICSPLINE"] = 3] = "CUBICSPLINE";
})(TreeAnimInterpType || (TreeAnimInterpType = {}));

// 材质类型
var MaterialType;
(function (MaterialType) {
    MaterialType["unlit"] = "unlit";
    MaterialType["pbr"] = "pbr";
    // 头发材质，在 pbr 材质基础上扩展
    MaterialType["hair"] = "hair";
})(MaterialType || (MaterialType = {}));
// 混合模式
var MaterialBlending;
(function (MaterialBlending) {
    MaterialBlending[MaterialBlending["opaque"] = 100] = "opaque";
    MaterialBlending[MaterialBlending["masked"] = 101] = "masked";
    MaterialBlending[MaterialBlending["translucent"] = 102] = "translucent";
    MaterialBlending[MaterialBlending["additive"] = 103] = "additive";
})(MaterialBlending || (MaterialBlending = {}));
// 透明度模式
var MaterialAlphaMode;
(function (MaterialAlphaMode) {
    MaterialAlphaMode["OPAQUE"] = "OPAQUE";
    MaterialAlphaMode["MASK"] = "MASK";
    MaterialAlphaMode["BLEND"] = "BLEND";
})(MaterialAlphaMode || (MaterialAlphaMode = {}));

/**
 * 3D渲染模式：将渲染过程中的中间结果输出，主要用于排查渲染效果问题，支持 pbr 和 unlit 材质
 */
var RenderMode3D;
(function (RenderMode3D) {
    /**
       * 正常渲染
       */
    RenderMode3D["none"] = "none";
    /**
       * 纹理坐标
       */
    RenderMode3D["uv"] = "uv";
    /**
       * 世界坐标法线
       */
    RenderMode3D["normal"] = "normal";
    /**
       * 基础颜色
       */
    RenderMode3D["basecolor"] = "basecolor";
    /**
       * 基础颜色 Alpha
       */
    RenderMode3D["alpha"] = "alpha";
    /**
       * 金属度
       */
    RenderMode3D["metallic"] = "metallic";
    /**
       * 粗超度
       */
    RenderMode3D["roughness"] = "roughness";
    /**
       * 环境遮蔽
       */
    RenderMode3D["ao"] = "ao";
    /**
       * 自发光
       */
    RenderMode3D["emissive"] = "emissive";
})(RenderMode3D || (RenderMode3D = {}));
/***
 * 一个 meshItem 打包出来的 json 案例：
 *
 * {
 *   type:'mesh',
 *   pluginName:'model',
 *   content:{
 *     options:{
 *       skins:{},
 *       primitives:[
 *         {
 *           material:{
 *           baseColorFactor:[0,0.5,1,1],
 *           baseColorTexture: {
 *              magFilter:33071,
 *              imageFileId:[0,556,1342]
 *            }
 *           },
 *           geometry:[0,0,556]
 *         }
 *       ]
 *     }
 *
 *   }
 * }
 * bins:[{url:'https://to/some/bin'}]
 */

/**
 * 如果二进制被压缩，js会解码数据，目前默认使用pako
 */
var BinaryCompressMethod;
(function (BinaryCompressMethod) {
    BinaryCompressMethod[BinaryCompressMethod["none"] = 0] = "none";
    BinaryCompressMethod[BinaryCompressMethod["pako"] = 1] = "pako";
})(BinaryCompressMethod || (BinaryCompressMethod = {}));
var BinaryEnv;
(function (BinaryEnv) {
    BinaryEnv["fileSystem"] = "fs";
    BinaryEnv["runtime"] = "runtime";
    BinaryEnv["json"] = "json";
})(BinaryEnv || (BinaryEnv = {}));

var CameraClipMode;
(function (CameraClipMode) {
    /**
       * 剪裁上下
       */
    CameraClipMode[CameraClipMode["portrait"] = CAMERA_CLIP_MODE_VERTICAL] = "portrait";
    /**
       * 剪裁左右
       */
    CameraClipMode[CameraClipMode["landscape"] = CAMERA_CLIP_MODE_NORMAL] = "landscape";
})(CameraClipMode || (CameraClipMode = {}));
/**
 * 结束行为
 */
var CompositionEndBehavior;
(function (CompositionEndBehavior) {
    /**
       * 销毁
       */
    CompositionEndBehavior[CompositionEndBehavior["destroy"] = END_BEHAVIOR_DESTROY] = "destroy";
    /**
       * 暂停
       */
    CompositionEndBehavior[CompositionEndBehavior["pause"] = END_BEHAVIOR_PAUSE] = "pause";
    /**
       * 重播
       */
    CompositionEndBehavior[CompositionEndBehavior["restart"] = END_BEHAVIOR_RESTART] = "restart";
    /**
       * 无限播放
       */
    CompositionEndBehavior[CompositionEndBehavior["forward"] = END_BEHAVIOR_FORWARD] = "forward";
    /**
       * 销毁并保留最后一帧
       */
    CompositionEndBehavior[CompositionEndBehavior["pause_destroy"] = END_BEHAVIOR_PAUSE_AND_DESTROY] = "pause_destroy";
})(CompositionEndBehavior || (CompositionEndBehavior = {}));

var TextOverflow;
(function (TextOverflow) {
    // default, display all text, it may decrease letter space size
    TextOverflow[TextOverflow["display"] = 0] = "display";
    // clip overflowed text
    TextOverflow[TextOverflow["clip"] = 1] = "clip";
    // use icon('..') to replace overflowed text
    TextOverflow[TextOverflow["ellipsis"] = 2] = "ellipsis";
})(TextOverflow || (TextOverflow = {}));
var TextAlignment;
(function (TextAlignment) {
    /**
     * text alignment starts from（x,y) to right direction
     * 从(x,y)开始第一个字符，向右边延伸
     */
    TextAlignment[TextAlignment["left"] = 0] = "left";
    /**
     * (x,y) is middle position of text, where (left + right)/2 =(x,y)
     * (x,y) 为文字中间位置，（最左位置 + 最右位置)/2 = (x,y)
     */
    TextAlignment[TextAlignment["middle"] = 1] = "middle";
    /**
     * text alignment ends with（x,y) from left direction
     * 从(x,y)结束最后一个字符，向左边延伸
     */
    TextAlignment[TextAlignment["right"] = 2] = "right";
})(TextAlignment || (TextAlignment = {}));
var FontStyle;
(function (FontStyle) {
    FontStyle[FontStyle["normal"] = 0] = "normal";
    FontStyle[FontStyle["italic"] = 1] = "italic";
    FontStyle[FontStyle["oblique"] = 2] = "oblique";
})(FontStyle || (FontStyle = {}));

/*********************************************/
/*               元素属性参数类型               */
/*********************************************/
/**
 * 渲染等级
 */
var RenderLevel;
(function (RenderLevel) {
    RenderLevel["S"] = "S";
    RenderLevel["APlus"] = "A+";
    RenderLevel["A"] = "A";
    RenderLevel["BPlus"] = "B+";
    RenderLevel["B"] = "B";
    RenderLevel["F"] = "F";
})(RenderLevel || (RenderLevel = {}));
/**
 * 混合模式
 */
var BlendingMode;
(function (BlendingMode) {
    /**
       * 普通混合模式
       */
    BlendingMode[BlendingMode["alpha"] = 0] = "alpha";
    /**
       * 叠加混合模式
       */
    BlendingMode[BlendingMode["ADD"] = 1] = "ADD";
    /**
       * 相乘混合模式
       */
    BlendingMode[BlendingMode["MULTIPLY"] = 2] = "MULTIPLY";
    /**
       * 亮度混合模式
       */
    BlendingMode[BlendingMode["BRIGHTNESS"] = 3] = "BRIGHTNESS";
    /**
       * 减色混合模式
       */
    BlendingMode[BlendingMode["SUBTRACTION"] = 4] = "SUBTRACTION";
    /**
       * 强光混合模式
       */
    BlendingMode[BlendingMode["STRONG_LIGHT"] = 5] = "STRONG_LIGHT";
    /**
       * 弱光混合模式
       */
    BlendingMode[BlendingMode["WEAK_LIGHT"] = 6] = "WEAK_LIGHT";
    /**
       * 亮度叠加混合模式
       */
    BlendingMode[BlendingMode["SUPERPOSITION"] = 7] = "SUPERPOSITION";
    /**
     * 卡片闪光模式
     */
    BlendingMode[BlendingMode["FLASH"] = 8] = "FLASH";
})(BlendingMode || (BlendingMode = {}));
/**
 * 单双面模式
 */
var SideMode;
(function (SideMode) {
    /**
       * 双面模式
       */
    SideMode[SideMode["DOUBLE"] = 1032] = "DOUBLE";
    /**
       * 正面模式
       */
    SideMode[SideMode["FRONT"] = 1028] = "FRONT";
    /**
       * 背面模式
       */
    SideMode[SideMode["BACK"] = 1029] = "BACK";
})(SideMode || (SideMode = {}));
/**
 * 蒙版模式
 */
var MaskMode;
(function (MaskMode) {
    /**
       * 无蒙版
       */
    MaskMode[MaskMode["NONE"] = 0] = "NONE";
    /**
       * 蒙版
       */
    MaskMode[MaskMode["MASK"] = 1] = "MASK";
    /**
       * 被遮挡
       */
    MaskMode[MaskMode["OBSCURED"] = 2] = "OBSCURED";
    /**
       * 被反向遮挡
       */
    MaskMode[MaskMode["REVERSE_OBSCURED"] = 3] = "REVERSE_OBSCURED";
})(MaskMode || (MaskMode = {}));
/**
 * 发射器形状
 */
var ShapeType;
(function (ShapeType) {
    /**
       * 没有类型
       */
    ShapeType[ShapeType["NONE"] = 0] = "NONE";
    /**
       * 圆球
       */
    ShapeType[ShapeType["SPHERE"] = 1] = "SPHERE";
    /**
       * 圆锥
       */
    ShapeType[ShapeType["CONE"] = 2] = "CONE";
    /**
       * 半球
       */
    ShapeType[ShapeType["HEMISPHERE"] = 3] = "HEMISPHERE";
    /**
       * 圆
       */
    ShapeType[ShapeType["CIRCLE"] = 4] = "CIRCLE";
    /**
       * 圆环
       */
    ShapeType[ShapeType["DONUT"] = 5] = "DONUT";
    /**
       * 矩形
       */
    ShapeType[ShapeType["RECTANGLE"] = 6] = "RECTANGLE";
    /**
       * 矩形框
       */
    ShapeType[ShapeType["RECTANGLE_EDGE"] = 7] = "RECTANGLE_EDGE";
    /**
       * 直线
       */
    ShapeType[ShapeType["EDGE"] = 8] = "EDGE";
    /**
       * 贴图
       */
    ShapeType[ShapeType["TEXTURE"] = 9] = "TEXTURE";
})(ShapeType || (ShapeType = {}));
/**
 * 插件类型
 */
var PluginType;
(function (PluginType) {
    /**
       * 陀螺仪
       */
    PluginType[PluginType["GYROSCOPE"] = 0] = "GYROSCOPE";
    /**
       * Spine
       */
    PluginType[PluginType["SPINE"] = 1] = "SPINE";
})(PluginType || (PluginType = {}));
/**
 * 交互类型
 */
var InteractType;
(function (InteractType) {
    /**
       * 点击
       */
    InteractType[InteractType["CLICK"] = 0] = "CLICK";
    /**
       * 消息,
       * 前端收到onMessageItem回调
       */
    InteractType[InteractType["MESSAGE"] = 1] = "MESSAGE";
    /**
       * 拖拽
       */
    InteractType[InteractType["DRAG"] = 2] = "DRAG";
    InteractType[InteractType["LOOP"] = 3] = "LOOP";
})(InteractType || (InteractType = {}));
/**
 * 交互行为
 */
var InteractBehavior;
(function (InteractBehavior) {
    /**
       * 无
       */
    InteractBehavior[InteractBehavior["NONE"] = 0] = "NONE";
    /**
       * 通知
       */
    InteractBehavior[InteractBehavior["NOTIFY"] = 1] = "NOTIFY";
    /**
       * 重置播放器
       */
    InteractBehavior[InteractBehavior["RESUME_PLAYER"] = 2] = "RESUME_PLAYER";
    /**
       * 清除元素
       */
    InteractBehavior[InteractBehavior["REMOVE"] = 3] = "REMOVE";
    /**
       * 暂停播放器
       */
    InteractBehavior[InteractBehavior["PAUSE"] = 4] = "PAUSE";
})(InteractBehavior || (InteractBehavior = {}));
/**
 * 元素类型
 */
var ItemType;
(function (ItemType) {
    /**
       * 错误元素
       */
    ItemType["base"] = "0";
    /**
       * 图层元素
       */
    ItemType["sprite"] = "1";
    /**
       * 粒子元素
       */
    ItemType["particle"] = "2";
    /**
       * 空节点元素
       */
    ItemType["null"] = "3";
    /**
       * 交互元素
       */
    ItemType["interact"] = "4";
    /**
       * 插件元素
       */
    ItemType["plugin"] = "5";
    /**
       * 相机元素
       */
    ItemType["camera"] = "6";
    /**
       * 预合成元素
       */
    ItemType["composition"] = "7";
    /**
       * 滤镜图层
       */
    ItemType["filter"] = "8";
    /**
     * text 元素
     */
    ItemType["text"] = "9";
    ItemType["tree"] = "tree";
    /**
     * Mesh 元素
     */
    ItemType["mesh"] = "mesh";
    ItemType["light"] = "light";
    /**
     * 天空盒元素
     */
    ItemType["skybox"] = "skybox";
    /**
       * Spine元素
       */
    ItemType["spine"] = "spi";
    ItemType["video"] = "av";
    ItemType["gyroscope"] = "gyro";
})(ItemType || (ItemType = {}));
/**
 * 渲染模式
 */
var RenderMode;
(function (RenderMode) {
    /**
       * 广告牌模式
       */
    RenderMode[RenderMode["BILLBOARD"] = 0] = "BILLBOARD";
    /**
       * 网格模式
       */
    RenderMode[RenderMode["MESH"] = 1] = "MESH";
    /**
       * 垂直广告牌模式
       */
    RenderMode[RenderMode["VERTICAL_BILLBOARD"] = 2] = "VERTICAL_BILLBOARD";
    /**
       * 水平广告牌模式
       */
    RenderMode[RenderMode["HORIZONTAL_BILLBOARD"] = 3] = "HORIZONTAL_BILLBOARD";
})(RenderMode || (RenderMode = {}));
/**
 * 变换中心
 */
var ParticleOrigin;
(function (ParticleOrigin) {
    /**
       * 水平和垂直中点
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_CENTER"] = 0] = "PARTICLE_ORIGIN_CENTER";
    /**
       * 水平左侧 垂直顶部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_LEFT_TOP"] = 1] = "PARTICLE_ORIGIN_LEFT_TOP";
    /**
       * 水平左侧 垂直中间
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_LEFT_CENTER"] = 2] = "PARTICLE_ORIGIN_LEFT_CENTER";
    /**
       * 水平左侧 垂直底部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_LEFT_BOTTOM"] = 3] = "PARTICLE_ORIGIN_LEFT_BOTTOM";
    /**
       * 水平中间 垂直顶部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_CENTER_TOP"] = 4] = "PARTICLE_ORIGIN_CENTER_TOP";
    /**
       * 水平中间 垂直底部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_CENTER_BOTTOM"] = 5] = "PARTICLE_ORIGIN_CENTER_BOTTOM";
    /**
       * 水平右侧 垂直顶部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_RIGHT_TOP"] = 6] = "PARTICLE_ORIGIN_RIGHT_TOP";
    /**
       * 水平右侧 垂直中间
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_RIGHT_CENTER"] = 7] = "PARTICLE_ORIGIN_RIGHT_CENTER";
    /**
       * 水平右侧 垂直底部
       */
    ParticleOrigin[ParticleOrigin["PARTICLE_ORIGIN_RIGHT_BOTTOM"] = 8] = "PARTICLE_ORIGIN_RIGHT_BOTTOM";
})(ParticleOrigin || (ParticleOrigin = {}));

/*********************************************/
/*               基本数值属性参数              */
var ValueType;
(function (ValueType) {
    /**
       * 常数
       */
    ValueType[ValueType["CONSTANT"] = 0] = "CONSTANT";
    /**
       * 二维常数向量
       */
    ValueType[ValueType["CONSTANT_VEC2"] = 1] = "CONSTANT_VEC2";
    /**
       * 三维常数向量
       */
    ValueType[ValueType["CONSTANT_VEC3"] = 2] = "CONSTANT_VEC3";
    /**
       * 四维常数向量
       */
    ValueType[ValueType["CONSTANT_VEC4"] = 3] = "CONSTANT_VEC4";
    /**
       * 随机数
       */
    ValueType[ValueType["RANDOM"] = 4] = "RANDOM";
    /**
       * 直线
       */
    ValueType[ValueType["LINE"] = 5] = "LINE";
    /**
       * 曲线
       */
    ValueType[ValueType["CURVE"] = 6] = "CURVE";
    /**
       * 贝塞尔路径
       */
    ValueType[ValueType["BEZIER_PATH"] = 7] = "BEZIER_PATH";
    /**
       * 颜色
       */
    ValueType[ValueType["RGBA_COLOR"] = 8] = "RGBA_COLOR";
    /**
       * 渐变色
       */
    ValueType[ValueType["GRADIENT_COLOR"] = 9] = "GRADIENT_COLOR";
    /**
       * 蒙版形状点集
       */
    ValueType[ValueType["SHAPE_POINTS"] = 10] = "SHAPE_POINTS";
    /**
       * 蒙版形状切分
       */
    ValueType[ValueType["SHAPE_SPLITS"] = 11] = "SHAPE_SPLITS";
    /**
       *直线路径
       */
    ValueType[ValueType["LINEAR_PATH"] = 12] = "LINEAR_PATH";
    /**
       * 多色
       */
    ValueType[ValueType["COLORS"] = 13] = "COLORS";
    ValueType[ValueType["BEZIER"] = 14] = "BEZIER";
    /**
       * 二进制指针
       */
    ValueType[ValueType["BINARY"] = 20] = "BINARY";
    ValueType[ValueType["BINARY_BEZIER"] = 21] = "BINARY_BEZIER";
    ValueType[ValueType["BINARY_BEZIER_PATH"] = 22] = "BINARY_BEZIER_PATH";
})(ValueType || (ValueType = {}));
var VideoFrameFormat;
(function (VideoFrameFormat) {
    VideoFrameFormat[VideoFrameFormat["RGB"] = 3] = "RGB";
    VideoFrameFormat[VideoFrameFormat["NV12"] = 1] = "NV12";
    VideoFrameFormat[VideoFrameFormat["I420"] = 2] = "I420";
    VideoFrameFormat[VideoFrameFormat["I420A"] = 4] = "I420A";
})(VideoFrameFormat || (VideoFrameFormat = {}));

const {
  PredyPlugin,
  registerPlugin,
  VFXItem
} = require('@predy-js/player');
class DrawingFramePlugin extends PredyPlugin {
  constructor() {
    super(...arguments);
    this._aspect = 1;
  }
  get aspect() {
    return this._aspect;
  }
  set aspect(v) {
    this._aspect = v;
    this.changeAspect(v);
  }
  onCompositionReset(composition, renderFrame) {
    const mesh = this.mesh = new renderInterface.Mesh({
      priority: 3000,
      material: {
        name: 'editor-frame',
        shader: {
          vertex,
          fragment,
          shared: true,
          cacheId: 'editor-frame'
        },
        states: {
          blending: false,
          cullFaceEnabled: false
        },
        dataBlocks: [{
          uniformValues: {
            uScale: [0.99, 0.99],
            uColor: [1, 0, 0]
          },
          uniformSemantics: {
            uEditorTransform: 'EDITOR_TRANSFORM'
          }
        }]
      },
      geometry: new renderInterface.Geometry({
        mode: renderInterface.constants.LINES,
        attributes: {
          aPoint: {
            size: 2,
            data: new Float32Array([-1, 1, 1, 1, 1, -1, -1, -1])
          }
        },
        index: {
          data: new Uint16Array([0, 1, 1, 2, 2, 3, 3, 0])
        }
      })
    });
    this.composition = composition;
    this.changeAspect(this.aspect);
    this._meshToAdd = mesh;
  }
  changeAspect(aspect) {
    let scale = [0.999, 0.999];
    const composition = this.composition;
    if (!composition) {
      return;
    }
    if (composition.env === "editor") {
      const p1 = composition.renderer.width / composition.renderer.height;
      const mode = composition.getCamera().params.clipMode;
      if (mode === CameraClipMode.landscape) {
        scale = [0.999 / p1 * aspect, 0.999];
      } else {
        scale = [0.999, p1 / aspect * 0.999];
      }
    }
    this.mesh.material.defaultDataBlock.setUniformValue('uScale', scale);
  }
  prepareRenderFrame(composition, frame) {
    if (this._meshToAdd) {
      let rp = frame.renderPasses.find(rp => rp.name === 'editor-gizmo');
      if (!rp) {
        rp = new renderInterface.RenderPass({
          priority: renderInterface.RenderPassPriorityPostprocess,
          name: 'editor-gizmo',
          meshes: [this._meshToAdd]
        });
      }
      frame.addRenderPass(rp);
      this._meshToAdd = undefined;
    }
    return false;
  }
}
const vertex = "\n  precision highp float;\n  attribute vec2 aPoint;\n  uniform vec4 uEditorTransform;\n  uniform vec2 uScale;\n  void main(){\n    vec2 point = aPoint * uScale * uEditorTransform.xy + uEditorTransform.zw; \n    gl_Position = vec4(point,0.,1.0);\n  }\n";
const fragment = "\n precision highp float;\n \n uniform vec3 uColor;\n void main(){\n  gl_FragColor = vec4(uColor,1.0);\n }\n\n";
function registerDrawingFramePlugin() {
  registerPlugin('drawing-frame', DrawingFramePlugin, VFXItem, true);
}

exports.DrawingFramePlugin = DrawingFramePlugin;
exports.registerDrawingFramePlugin = registerDrawingFramePlugin;
