// @predy-js/player/dist/serialize.js
import { ItemType, ValueType } from '@predy-js/specification';

if (typeof __PredyConfig__ === 'undefined') {
  if (typeof window === 'object') {
    //@ts-expect-error
    window.__PredyConfig__ = {};
  } else {
    //@ts-expect-error
    global.__PredyConfig__ = {};
  }
}
__PredyConfig__;

/**
 * Created by brian on 20/07/2017.
 */
function arrAdd(arr, item) {
  if (!arr.includes(item)) {
    arr.push(item);
    return true;
  }
  return false;
}
const hasOwnProperty = Object.prototype.hasOwnProperty;
function forEach(object, callback, thisObj) {
  if (isObj(object)) {
    if (thisObj === undefined) {
      thisObj = object;
    }
    if (object instanceof Array || 'length' in object) {
      const len = object.length;
      for (let i = 0; i < len; i++) {
        callback.call(thisObj, object[i], i);
      }
    } else {
      for (const name in object) {
        if (hasOwnProperty.call(object, name)) {
          callback.call(thisObj, object[name], name);
        }
      }
    }
  }
  return object;
}
function isObj(any) {
  return typeof any === 'object' && any;
}

function getCompositionGraph(comp) {
  const childrenMap = {};
  const nodeMap = {};
  const topNodes = [];
  const treeNodesMap = {};
  comp.items.forEach(item => collectNodes(item));
  function collectNodes(item, treeNodeChildren) {
    const node = {
      name: item.name,
      type: item.type,
      id: item.id,
      transform: item.transform,
      parentId: item.parentId
    };
    const pid = node.parentId;
    if (pid) {
      if (!childrenMap[pid]) {
        childrenMap[pid] = [];
      }
      childrenMap[pid].push(node);
    } else {
      topNodes.push(node);
    }
    nodeMap[node.id] = node;
    if (node.type === ItemType.tree) {
      var _children;
      let children;
      let nodes;
      const isTreeNode = node.id.includes('^');
      if (isTreeNode) {
        nodes = treeNodesMap[node.id.slice(0, node.id.indexOf('^'))];
        children = treeNodeChildren;
        node.subType = 'node';
        node.parentId = replaceTreeParentId(node.parentId);
      } else {
        nodes = item.tree.nodes;
        children = item.tree.children;
        treeNodesMap[item.id] = nodes;
      }
      (_children = children) == null || _children.forEach(index => {
        const tn = nodes[index];
        collectNodes({
          name: tn.name,
          type: 'tree',
          id: replaceTreeParentId(item.id) + "^" + (tn.id || index),
          transform: tn.transform,
          parentId: item.id
        }, tn.children);
      });
    }
  }
  Object.values(nodeMap).forEach(node => {
    const children = childrenMap[node.id];
    if (children) {
      node.children = children;
    }
  });
  return {
    name: comp.name,
    id: comp.id,
    nodes: topNodes
  };
}
function replaceTreeParentId(str) {
  const idx = str.indexOf('^');
  return idx > -1 ? str.substring(0, idx) : str;
}

const TextureSourceTypeImage = 2;
class Serializer {
  static serializeTexturesAsync(textures) {
    return new Promise(function (resolve, reject) {
      const srcSet = new Map();
      const files = [];
      const imageMap = new Map();
      const cubes = [];
      const cubeMap = new Map();
      for (let i = 0; i < textures.length; i++) {
        const texture = textures[i];
        const sourceFrom = texture.sourceFrom;
        if (sourceFrom && sourceFrom.type === TextureSourceTypeImage && sourceFrom.target !== WebGLRenderingContext.TEXTURE_CUBE_MAP) {
          files.push(fetch(sourceFrom.url).then(res => res.arrayBuffer()).then(buffer => {
            imageMap.set(texture, buffer);
            return buffer;
          }));
          continue;
        }
        if (texture.target === WebGLRenderingContext.TEXTURE_CUBE_MAP) {
          cubes.push(loadMipmapsAsync(texture, cubes.length).then(result => cubeMap.set(texture, result)));
          continue;
        }
        const image = texture.image;
        if (instanceofImageBitmap(image)) {
          files.push(getImageFileContent(image).then(buffer => {
            imageMap.set(texture, buffer);
            return buffer;
          }));
        } else if (instanceofHTMLImageElement(image)) {
          if (!srcSet.has(image.src)) {
            srcSet.set(image.src, getImageFileContent(image));
          }
          files.push(srcSet.get(image.src).then(buffer => {
            imageMap.set(texture, buffer);
            return buffer;
          }));
        } else {
          return reject(Error("tex " + i + " image should be HTMLImage or HTMLCanvas"));
        }
      }
      Promise.all(cubes).then(function () {
        // eslint-disable-next-line promise/no-nesting
        Promise.all(files).then(function (files) {
          const images = Array.from(new Set(files));
          const retTextures = [];
          let tex;
          const bins = [];
          for (let i = 0; i < textures.length; i++) {
            const texture = textures[i];
            if (texture.target === WebGLRenderingContext.TEXTURE_CUBE_MAP) {
              const info = cubeMap.get(texture);
              bins.push(info.data);
              tex = Object.assign({}, texture, {
                mipmaps: info.mipmaps,
                sourceType: 7
              });
              delete tex.cube;
            } else {
              const arrayBuffer = imageMap.get(texture);
              tex = Object.assign({}, texture, {
                source: images.indexOf(arrayBuffer)
              });
              delete tex.mipmaps;
            }
            delete tex.sourceFrom;
            delete tex.image;
            retTextures.push(tex);
          }
          resolve({
            version: '1.0',
            images,
            bins,
            textures: retTextures
          });
        }, reject);
      }, reject);
    });
  }
  static deserializeGeometry(geo, data) {
    const ret = Object.assign({}, geo);
    if (ret.index) {
      ret.index = Object.assign({}, ret.index);
      if (ret.index.data) {
        var _geo$index;
        ret.index.data = typedArrayFromBinary(data, (_geo$index = geo.index) == null ? void 0 : _geo$index.data);
      }
    }
    forEach(geo.attributes, function (attr, name) {
      if (attr.data) {
        ret.attributes[name] = Object.assign({}, attr, {
          data: typedArrayFromBinary(data, attr.data)
        });
      } else {
        ret.attributes[name] = Object.assign({}, attr, {
          data: undefined
        });
      }
    });
    return ret;
  }
  static concatArrayBuffers(buffers, pointers, overwrite) {
    const bufferInfo = new Map();
    let offset = 0;
    for (let i = 0; i < buffers.length; i++) {
      const buffer = buffers[i];
      const content = [0, offset, buffer.byteLength];
      const type = getBinaryType(buffer);
      if (type) {
        content[3] = type;
      }
      bufferInfo.set(buffer, content);
      offset += buffer.byteLength;
      //padding 4bytes for f32
      offset = padding4(offset);
    }
    if (!pointers) {
      pointers = buffers.map(function (buffer, i) {
        return [ValueType.BINARY, [i, 0, buffer.byteLength]];
      });
    }
    const retBuffer = concatBuffers(buffers, bufferInfo, offset);
    const mappedPointers = pointers.map(function (pointer) {
      const content = pointer[1];
      const buffer = buffers[content[0]];
      if (!buffer) {
        throw Error("buffer index " + content[0] + " not found");
      }
      const originStart = content[1] || 0;
      const byteLength = content[2] || buffer.byteLength - originStart;
      const info = bufferInfo.get(buffer);
      const rc = [0, originStart + info[1], byteLength];
      if (content[3]) {
        rc[3] = content[3];
      } else if (info[3]) {
        rc[3] = info[3];
      }
      if (overwrite) {
        pointer[1] = rc;
        return pointer;
      }
      return [ValueType.BINARY, rc];
    });
    return {
      data: retBuffer,
      pointers: mappedPointers
    };
  }
  static serializeGeometries(geometries, bufferIndex) {
    if (bufferIndex === void 0) {
      bufferIndex = 0;
    }
    const buffers = [];
    const bufferInfo = new Map();
    let offset = 0;
    geometries.forEach(options => {
      var _options$index;
      addBuffer((_options$index = options.index) == null ? void 0 : _options$index.data);
      forEach(options.attributes, function (attr) {
        addBuffer(attr.data);
      });
    });
    const retBuffer = concatBuffers(buffers, bufferInfo, offset);
    const mappedGeos = geometries.map(options => {
      const ret = Object.assign({}, options);
      if (ret.index) {
        ret.index = replaceAttribute(ret.index);
      }
      ret.attributes = {};
      forEach(options.attributes, function (attr, name) {
        ret.attributes[name] = replaceAttribute(attr);
      });
      return ret;
    });
    return {
      data: retBuffer,
      geometries: mappedGeos,
      version: '1.0'
    };
    function addBuffer(buffer) {
      if (buffer && arrAdd(buffers, buffer)) {
        const bufferType = getBinaryType(buffer);
        const content = [bufferIndex, offset, buffer.byteLength];
        if (bufferType) {
          content.push(bufferType);
        }
        bufferInfo.set(buffer, content);
        offset += buffer.byteLength;
        //padding 4bytes for f32
        offset = padding4(offset);
      }
    }
    function replaceAttribute(attr) {
      let source;
      if (attr.data) {
        const info = bufferInfo.get(attr.data);
        if (info) {
          source = {
            data: [ValueType.BINARY, info]
          };
        }
      }
      return Object.assign({}, attr, source);
    }
  }
  static typedArrayFromBinary(binary, pointer) {
    return typedArrayFromBinary(binary, pointer);
  }
}
function concatBuffers(buffers, bufferInfo, totalSize) {
  const ret = new Uint8Array(totalSize);
  buffers.forEach(buffer => {
    const info = bufferInfo.get(buffer);
    const source = buffer instanceof ArrayBuffer ? new Uint8Array(buffer, 0, info[2]) : new Uint8Array(buffer.buffer, buffer.byteOffset, info[2]);
    ret.set(source, info[1]);
  });
  return ret.buffer;
}
function typedArrayFromBinary(binary, pointer) {
  const [index, start, byteLength, type] = pointer[1];
  if (type === 'b') {
    return binary[index].slice(start, start + byteLength);
  }
  const CTRL = ctrlMap[type] || Float32Array;
  return new CTRL(binary[index], start, byteLength / CTRL.BYTES_PER_ELEMENT);
}
function loadMipmapsAsync(tex, dataIndex) {
  const mipmaps = tex.mipmaps || [tex.cube];
  return Promise.all(mipmaps.map(mipmap => Promise.all(mipmap.map(function (imageLike) {
    if (instanceofImageBitmap(imageLike) || instanceofHTMLImageElement(imageLike)) {
      return getImageFileContent(imageLike);
    }
    return Promise.reject(Error('invalid image format'));
  })))).then(function (mipmaps) {
    const images = [];
    mipmaps.forEach(mipmap => images.push(...mipmap));
    const result = Serializer.concatArrayBuffers(images);
    const cubeMipmaps = [];
    for (let i = 0; i < mipmaps.length; i++) {
      const mipmap = mipmaps[i];
      const mipmapPointers = [];
      for (let j = 0; j < mipmap.length; j++) {
        const p = mipmapPointers[j] = result.pointers[i * 6 + j];
        p[1][0] = dataIndex;
        p[1].length = 3;
      }
      cubeMipmaps.push(mipmapPointers);
    }
    return {
      data: result.data,
      mipmaps: cubeMipmaps
    };
  });
}
function getImageFileContent(image) {
  const cvs = document.createElement('canvas');
  cvs.width = image.width;
  cvs.height = image.height;
  cvs.getContext('2d').drawImage(image, 0, 0);
  return new Promise(function (resolve, reject) {
    cvs.toBlob(function (b) {
      if (b) {
        resolve(b.arrayBuffer());
      } else {
        reject(Error('no canvas blob'));
      }
    }, 'image/webp', 1);
  });
}
function getBinaryType(arr) {
  if (arr instanceof Uint8Array) {
    return 'u8';
  } else if (arr instanceof Int8Array) {
    return 'i8';
  } else if (arr instanceof Uint16Array) {
    return 'u16';
  } else if (arr instanceof Int16Array) {
    return 'i16';
  } else if (arr instanceof Uint32Array) {
    return 'u32';
  } else if (arr instanceof Int32Array) {
    return 'i32';
  } else if (arr instanceof Float32Array) {
    return '';
  }
  return 'b';
}
const ctrlMap = {
  i8: Int8Array,
  u8: Uint8Array,
  i16: Int16Array,
  u16: Uint16Array,
  f32: Float32Array,
  i32: Int32Array,
  u32: Uint32Array
};
function padding4(n) {
  return Math.ceil(n / 4) * 4;
}

export { Serializer, getBinaryType, getCompositionGraph };
